### PR TITLE
feat: scroll account

### PR DESCRIPTION
**Pull request by greged93**
Propagates all changes made on accounts from https://github.com/scroll-tech/reth/pull/33. This PR should only be reviewed once https://github.com/scroll-tech/reth/pull/33 is merged.

Adds `test_create` in the execution stage in order to test the runtime computation of the Scroll fields. The caching characteristic is already covered by `test_selfdestruct`.

Builds towards #7

### CODE REVIEW

File: crates/payload/basic/src/lib.rs [hunk 1]
```diff
@@ -204,7 +204,12 @@ where
                 // this only includes changed accounts and storage but is better than nothing
                 let storage =
                     acc.storage.iter().map(|(key, slot)| (*key, slot.present_value)).collect();
-                cached.insert_account(addr, info, storage);
+                cached.insert_account(
+                    addr,
+                    #[allow(clippy::useless_conversion)]
+                    info.into(),
```
Comment (greged93) [inline crates/payload/basic/src/lib.rs:210]: this conversion causes a loss of information since we convert a `ScrollAccountInfo` to an `AccountInfo`. However, this function is used for payload building in [Ethereum](https://github.com/paradigmxyz/reth/blob/4da6402ea33f931961e3bf54701d27de4fdbe298/crates/ethereum/payload/src/lib.rs#L163-L169). 

When payload building is implemented for Scroll, we might need to udpate the `CachedReads` to provide a `ScrollAccountInfo`. Because Milestone 1 doesn't require payload building, we leave this out for now.

File: crates/revm/src/database.rs [hunk 2]
```diff
@@ -99,6 +95,15 @@ impl<DB> DerefMut for StateProviderDatabase<DB> {
     }
 }
 
+#[cfg(any(not(feature = "scroll"), feature = "test-utils"))]
```
Comment (greged93) [inline crates/revm/src/database.rs]: we feature gate the implementation to be sure the `StateProviderDatabase` instead used anywhere in production code instead of the `ScrollStateProviderDatabase`.

File: crates/scroll/revm/src/states/bundle.rs [hunk 3]
```diff
@@ -70,9 +70,19 @@ impl From<(BundleState, &ScrollPostExecutionContext)> for ScrollBundleState {
     }
 }
 
+// This conversion can cause a loss of information since performed without additional context.
+#[cfg(any(not(feature = "scroll"), feature = "test-utils"))]
```
Comment (greged93) [inline crates/scroll/revm/src/states/bundle.rs:74]: we feature gate this lossful conversion to avoid using it in production code with the Scroll feature

File: crates/trie/trie/src/state.rs [hunk 4]
```diff
@@ -50,33 +50,6 @@ impl HashedPostState {
         Self { accounts, storages }
     }
 
-    /// Initialize [`HashedPostState`] from cached state.
-    /// Hashes all changed accounts and storage entries that are currently stored in cache.
-    pub fn from_cache_state<'a>(
```
Comment (greged93) [inline crates/trie/trie/src/state.rs]: this function is never used in the codebase and is only part of a public API. Removed in order to avoid having to maintain unused code.

File: crates/blockchain-tree/Cargo.toml [hunk 5]
```diff
@@ -76,12 +80,27 @@ test-utils = [
 	"reth-db-api/test-utils",
 	"reth-provider/test-utils",
 	"reth-trie-db/test-utils",
-	"reth-trie/test-utils"
+    "reth-trie/test-utils",
+    "revm/test-utils"
```
Comment (frisitano) [inline crates/blockchain-tree/Cargo.toml]: Nit: align dependencies.

File: crates/blockchain-tree/Cargo.toml [hunk 6]
```diff
@@ -76,12 +80,27 @@ test-utils = [
...
+    "reth-primitives/scroll",
+    "reth-revm/scroll",
+    "reth-execution-types/scroll",
+    "reth-db/scroll",
+    "reth-provider/scroll",
+    "reth-trie-db/scroll",
+    "reth-testing-utils/scroll",
+    "reth-trie/scroll",
+    "revm/scroll",
+	"reth-scroll-storage/scroll"
```
Comment (frisitano) [inline crates/blockchain-tree/Cargo.toml]: nit: let's use consistent indentation for dependencies.

File: crates/engine/util/src/reorg.rs [hunk 7]
```diff
@@ -377,13 +378,13 @@ where
     // and 4788 contract call
     state.merge_transitions(BundleRetention::PlainState);
 
-    let outcome: ExecutionOutcome = ExecutionOutcome::new(
-        state.take_bundle(),
-        Receipts::from(vec![receipts]),
+    let outcome = ExecutionOutcome::new(
+        state.finalize(),
+        Receipts::<Receipt>::from(vec![receipts]),
```
Comment (frisitano) [inline crates/engine/util/src/reorg.rs:383]: Is there a reason we added the generic `Receipt` definition here?

File: crates/revm/Cargo.toml [hunk 8]
```diff
@@ -52,12 +55,14 @@ test-utils = [
 	"revm/test-utils",
 	"reth-prune-types/test-utils",
     "reth-primitives-traits/test-utils",
+	"reth-scroll-execution/test-utils"
```
Comment (frisitano) [inline crates/revm/Cargo.toml]: nit: dependency alignment

File: crates/revm/Cargo.toml [hunk 9]
```diff
@@ -52,12 +55,14 @@ test-utils = [
...
 ]
 serde = [
     "revm/serde",
     "reth-trie?/serde",
     "alloy-eips/serde",
     "alloy-primitives/serde",
     "alloy-consensus/serde",
-    "reth-primitives-traits/serde",
+	"reth-primitives-traits/serde",
```
Comment (frisitano) [inline crates/revm/Cargo.toml]: nit: dependency alignment

File: crates/scroll/execution/Cargo.toml [hunk 10]
```diff
@@ -13,23 +13,17 @@ workspace = true
...
 # misc
 auto_impl.workspace = true
 
 [features]
 scroll = [
-	"reth-scroll-primitives",
-	"reth-scroll-revm/scroll",
-	"reth-scroll-storage/scroll",
+    "reth-scroll-primitives",
+    "revm/scroll",
```
Comment (frisitano) [inline crates/scroll/execution/Cargo.toml]: nit: dependency indentation

File: crates/scroll/execution/src/context.rs [hunk 11]
```diff
@@ -1,22 +1,26 @@
...
-use reth_revm::{database::StateProviderDatabase, db::CacheDB};
-use reth_revm::{Database, State};
+use revm::db::CacheDB;
+use revm::{Database, State};
 
 /// Finalize the execution of the type and return the output
 pub trait FinalizeExecution<Output> {
     /// Finalize the state and return the output.
     fn finalize(&mut self) -> Output;
 }
```
Comment (frisitano) [inline crates/scroll/execution/src/context.rs]: I suspect that this implementation in which we currently have three traits `FinalizeExecution`, `ContextFul` and `WithContext` is a little more complicated then it needs to be. I know we had discussed this previously but I thnk we could just have 
```rust
/// Finalize the execution of the type and return the output
pub trait FinalizeExecution {
    type Output;
    /// Finalize the state and return the output.
    fn finalize(&mut self) -> Self::Output;
}
```

and then bound in the strategy with `State<DB>: FinalizeExecution`. We would then implement `FinalizeExecution` for `State<StateProviderDatabase<DB>>` and `State<ScrollStateProviderDatabase<DB>>`. What is the reason we can't do this, I remember. you mentioned there was one.

File: crates/scroll/storage/Cargo.toml [hunk 12]
```diff
@@ -31,6 +32,7 @@ reth-revm = { workspace = true, features = ["test-utils"] }
 
 [features]
 scroll = [
-	"reth-primitives-traits/scroll",
-	"reth-scroll-revm/scroll",
+    "reth-scroll-execution/scroll",
+    "reth-primitives-traits/scroll",
+    "reth-scroll-revm/scroll",
```
Comment (frisitano) [inline crates/scroll/storage/Cargo.toml]: nit: indentation

File: crates/trie/common/Cargo.toml [hunk 13]
```diff
@@ -20,8 +20,9 @@ alloy-rlp = { workspace = true, features = ["arrayvec"] }
 alloy-trie = { workspace = true, features = ["serde"] }
 alloy-consensus.workspace = true
 alloy-genesis.workspace = true
-revm-primitives.workspace = true
 
+# revm-primitives scroll re-export
+revm-primitives = { package = "reth-scroll-revm", path = "../../scroll/revm" }
```
Comment (frisitano) [inline crates/trie/common/Cargo.toml:25]: Should we add this to the workspace `Cargo.toml`?

File: crates/trie/common/Cargo.toml [hunk 14]
```diff
@@ -51,7 +52,8 @@ test-utils = [
 	"dep:hash-db",
 	"arbitrary",
 	"reth-primitives-traits/test-utils",
-	"reth-codecs/test-utils"
+    "reth-codecs/test-utils",
+    "revm-primitives/test-utils"
```
Comment (frisitano) [inline crates/trie/common/Cargo.toml]: nit: indentation

File: crates/trie/trie/src/state.rs [hunk 15]
```diff
@@ -25,11 +25,11 @@ impl HashedPostState {
...
     /// state.
-    pub fn from_bundle_state<'a>(
-        state: impl IntoParallelIterator<Item = (&'a Address, &'a BundleAccount)>,
-    ) -> Self {
+    pub fn from_bundle_state(state: &BundleState) -> Self {
         let hashed = state
-            .into_par_iter()
+            .state
+            .iter()
+            .par_bridge()
```
Comment (frisitano) [inline crates/trie/trie/src/state.rs]: Is there a specific reason you have changed this implementation? This has implications for the state root work I'm working on. I think we should avoid changes to core data structures unless they are necessary.

File: crates/trie/trie/src/state.rs (hunk 4 above)
Comment (frisitano) [reply crates/trie/trie/src/state.rs]: I think we should keep this in to keep diff with upstream minimal. What overhead are we looking at in regards to code maintenance?

File: crates/trie/trie/src/state.rs [hunk 16]
```diff
@@ -459,49 +433,13 @@ mod tests {
...
-            account: Some(PlainAccount { info: account_info.clone(), storage }),
-            status: AccountStatus::Changed,
-        };
-
-        // Create a vector of tuples representing the cache state.
-        let state = vec![(&address, &account)];
-
-        // Convert the cache state into a hashed post state.
-        let hashed_state = HashedPostState::from_cache_state(state);
+        let hashed_state = HashedPostState::from_bundle_state(&state);
```
Comment (frisitano) [inline crates/trie/trie/src/state.rs]: If we revert `from_bundle_state` and reintroduce `from_cache_state` then we should revert these changes.

**Review by frisitano — COMMENTED** [review]
Looks good! I've left some comments in line. Regarding indentation of dependencies lets try and stay consistent with the formatting currently used in reth. Looks like there's a few `Cargo.toml` files in which we diverge. I also think we should not change the `HashedPostState` type. I would like to understand your reasoning for doing so. Also added comment around the pattern for finalization of the `State<DB>` I suspect we can simplify this but wanted to see from your perspective if there's a blocker to the proposed solution.

File: crates/blockchain-tree/Cargo.toml (hunk 5 above)
Comment (greged93) [reply crates/blockchain-tree/Cargo.toml]: oh this is strange, it appears aligned in my IDE and didn't notice this when pushing... will fix them all

File: crates/engine/util/src/reorg.rs (hunk 7 above)
Comment (greged93) [reply crates/engine/util/src/reorg.rs:383]: for some reason when I added the `finalize()`, the compiler wasn't able to infer the generic for `Receipts` anymore.

File: crates/scroll/execution/src/context.rs (hunk 11 above)
Comment (greged93) [reply crates/scroll/execution/src/context.rs]: The main issue with `State<DB>: FinalizeExecution` is that you need a different implementation of the `FinalizeExecution` depending on the database. You would need the following: 
- `impl<DB> FinalizeExecution for State<DB>` (blanket)
- `impl<DB> FinalizeExecution for State<StateProviderDatabase<DB>> `
- `impl<DB> FinalizeExecution for State<ScrollStateProviderDatabase<DB>>`

Each implementation would have a different `Output` concrete.

This isn't currently possible in Rust because the language doesn't provide [`specialization`](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md).

File: crates/trie/common/Cargo.toml (hunk 13 above)
Comment (greged93) [reply crates/trie/common/Cargo.toml:25]: I encountered an issue when adding both `revm` and `revm-primitives` to the workspace: any crate that imports both of them faces a compile issue due to importing twice the same crate. I could update them in order to just use `revm` but I felt like this would increase the diff...

File: crates/trie/trie/src/state.rs (hunk 15 above)
Comment (greged93) [reply crates/trie/trie/src/state.rs]: I think this is a slip up, I can't remember why I changed it and it's not necessary anymore, reverting.

File: crates/trie/trie/src/state.rs (hunk 4 above)
Comment (greged93) [reply crates/trie/trie/src/state.rs]: we need to add an implementation of `AccountInfo` to `Account` under the `scroll` feature. I am a little worried of having this implementation and missing a shadowing of `AccountInfo` by `ScrollAccountInfo` combined with an `into()` which could be painful to debug.

I could implement a specific function `into_account(info: AccountInfo)` also, in order to avoid any mix up. What do you think?

Comment (greged93) [reply crates/trie/trie/src/state.rs]: I pushed the revert of the deletion in a377aba, let me know if you want to keep it as is

File: crates/revm/Cargo.toml (hunk 8 above)
Comment (greged93) [reply crates/revm/Cargo.toml]: Fixed the alignment, but it looks like Reth doesn't use consistent alignment

File: crates/scroll/execution/src/context.rs (hunk 11 above)
Comment (frisitano) [reply crates/scroll/execution/src/context.rs]: Where does the requirement for the blanket implementation come from? From my observation, it looks like the executor is only ever instantiated with `StateProviderDatabase` and `CacheDb` (for tests). So we should be able to implement the trait for:

- `State<StateProviderDatabase<DB>>`
- `State<ScrollStateProviderDatabase<DB>>`
- `State<CacheDb<DB>>` (behind `test-utils` feature)

Comment (greged93) [reply crates/scroll/execution/src/context.rs]: You might be right, let me try this out and will push an update

Comment (greged93) [reply crates/scroll/execution/src/context.rs]: @frisitano I don't know why I thought this didn't work, it looks fine. The only remaining issue is a cyclic dep between `reth-revm` and `reth-scroll-execution`: `reth-scroll-execution` needs to import `reth-revm` in order to implement  the trait on `State<StateProviderDatabase<DB>>`, but `reth-revm` also has a function `record_executed_state<DB>(&mut self, statedb: &State<DB>)` which would require `State<DB>: FinalizeExecution`.

Trying to solve this, will update when I find a solution

File: crates/engine/util/src/reorg.rs (hunk 7 above)
Comment (frisitano) [reply crates/engine/util/src/reorg.rs:383]: Let's review this when we implement the new pattern for the `FinalizeExecution` trait. It may be that the change provides the compiler the additional context it needs.

File: crates/scroll/execution/src/context.rs (hunk 11 above)
Comment (frisitano) [reply crates/scroll/execution/src/context.rs]: Can you implement `FinalizeExecution` on `State<StateProviderDatabase<DB>>` in `reth-revm` instead?

File: crates/trie/trie/src/state.rs (hunk 4 above)
Comment (frisitano) [reply crates/trie/trie/src/state.rs]: This looks fine to me

File: crates/revm/Cargo.toml (hunk 8 above)
Comment (frisitano) [reply crates/revm/Cargo.toml]: ahh I see, let's just be internally consistent within each `Cargo.toml`.

File: crates/engine/util/src/reorg.rs (hunk 7 above)
Comment (greged93) [reply crates/engine/util/src/reorg.rs:383]: I updated the PR with the new pattern for `FinalizeExecution`, but unfortunately the compiler still needs this hint on the `Receipt` type.

File: crates/revm/src/witness.rs [hunk 17]
```diff
@@ -22,7 +22,11 @@ pub struct ExecutionWitnessRecord {
 
 impl ExecutionWitnessRecord {
     /// Records the state after execution.
-    pub fn record_executed_state<DB>(&mut self, statedb: &State<DB>) {
+    pub fn record_executed_state<DB>(
+        &mut self,
+        statedb: &State<DB>,
+        #[cfg(feature = "scroll")] context: &reth_scroll_primitives::ScrollPostExecutionContext,
```
Comment (greged93) [inline crates/revm/src/witness.rs:28]: this is required because of a cyclic dep: `reth-scroll-execution` needs to import `reth-revm`, to implement `State<StateProviderDatabase<DB>>`, and the trait can't be implemented elsewhere because of the orphan rule. This means `reth-revm` can't import `reth-scroll-execution`, so we can't have a bound on `State<DB>`.

File: crates/scroll/execution/src/context.rs (hunk 11 above)
Comment (greged93) [reply crates/scroll/execution/src/context.rs]: I commented on the PR but I can't because of the orphan rule

File: examples/custom-engine-types/src/main.rs [hunk 18]
```diff
@@ -1,21 +1,24 @@
...
+//! `PayloadAttributes` can be implemented for payload attributes types that are used as
 //! arguments to the `engine_forkchoiceUpdated` method. This type should be used to define and
 //! _spawn_ payload jobs.
 //!
-//! [PayloadBuilderAttributes] can be implemented for payload attributes types that _describe_
+//! `PayloadBuilderAttributes` can be implemented for payload attributes types that _describe_
 //! running payload jobs.
 //!
-//! Once traits are implemented and custom types are defined, the [EngineTypes] trait can be
+//! Once traits are implemented and custom types are defined, the `EngineTypes` trait can be
```
Comment (frisitano) [inline examples/custom-engine-types/src/main.rs:15]: Why are we changing the `[` to ` here?

**Review by frisitano — APPROVED** [review]
Looks good, left one comment inline in regards to doc comment change which I think we should revert unless there is good reason not to.

File: crates/scroll/execution/src/context.rs (hunk 11 above)
Comment (frisitano) [reply crates/scroll/execution/src/context.rs]: I'm a little confused about reference to the orphan rule. Rust allows you to implement a foreign trait on a local type. `StateproviderDatabase` is a local type in `reth-revm` and `FinalizeExecution` is the foreign trait. This should be allowed.

Comment (greged93) [reply crates/scroll/execution/src/context.rs]: I missed this as well, but you are implementing in on `State<StateproviderDatabase<DB>>`, not `StateproviderDatabase<DB>`, and `State` originates from `revm`.

File: examples/custom-engine-types/src/main.rs (hunk 18 above)
Comment (greged93) [reply examples/custom-engine-types/src/main.rs:15]: I honestly don't know why, but running `cargo docs --document-private-items` fails on this crate. It looks like it compiles it by default with the `scroll` feature, hence not including the imports which then causes `cargo docs` to issue a warning.

I tried exploring the imports with `cargo tree -p example-custom-engine-types -e features` to see if perhaps a crate downstream had the `scroll` feature active but didn't find any.

Comment (frisitano) [reply examples/custom-engine-types/src/main.rs:15]: Ok it's fine for now not a big issue.

**Comment by frisitano** [comment]
LGTM
